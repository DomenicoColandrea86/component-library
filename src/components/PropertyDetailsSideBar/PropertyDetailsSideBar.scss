.PropertyDetailsSideBar {
	$font-size-base: 12px;
	$font-family-base: 'Roboto-Regular', 'Helvetica Neue', Helvetica, Arial,
		sans-serif;
	$navbar-height: 70px;

	$blue: #2567ab;
	$darkBlue: #143e75;
	$glacier: #f0f2f4;
	$width: 450px;
	$padding: 20px;
	$header-font: 'Roboto Condensed-Bold', $font-family-base;
	$bold-font: 'Roboto-Bold', $font-family-base;
	$reg-font: $font-family-base;
	$toggle-control-height: 10px;

	top: $navbar-height;
	right: -$width;
	position: fixed;
	z-index: 200;
	width: $width;
	height: 100%;
	min-height: 100%;
	padding: 0 0 0 $padding;
	background: $glacier;
	box-shadow: none;
	transition: all linear 0.2s;

	&--is-open {
		right: 0;
		box-shadow: 5px 0 10px 0 rgba(0, 0, 0, 0.6);
	}

	&__close {
		cursor: pointer;
		position: absolute;
		left: 5px;
		top: calc(50% - #{$navbar-height});
		background: transparent;
		outline: none;
		padding: 0;
		width: 0;
		height: 0;
		border: 0 solid transparent;
		border-bottom-width: $toggle-control-height;
		border-top-width: $toggle-control-height;
		border-left: solid $toggle-control-height $darkBlue;
	}

	&__content {
		height: 100%;
		min-height: 100%;
		padding: $padding;
		background: white;
	}

	&__header {
		margin: 0 0 $padding 0;
		color: $blue;
		font-size: 18px;
		font-family: $header-font;
	}

	&__btn-group {
		border: 1px solid $blue;
		overflow: hidden;
		border-radius: 3px;
		display: flex;
	}

	&__btn {
		padding: 0.4em;
		border: solid 1px $blue;
		border-radius: 3px;
		display: inline-block;
		text-align: center;
		flex-grow: 1;
		flex-basis: 100%;
		border-radius: 0;
		border: none;
		border-right: 1px solid $blue;
		font-size: 10px;
		font-family: $bold-font;
		text-transform: uppercase;

		&:hover,
		&:focus {
			text-decoration: none;
		}

		&:not(.PropertyDetailsSideBar__btn--disabled) {
			&:hover,
			&:focus {
				cursor: pointer;
				color: white;
				background-color: $blue;
			}
		}

		&:last-child {
			border-right: none;
		}

		&--disabled {
			cursor: not-allowed;
			color: #767676;
			background: #f4f7f9;
		}

		&--active {
			color: white;
			background: $blue;
		}
	}
	&__comments {
		margin-top: $padding;
		font-family: $reg-font;
		font-size: $font-size-base;
	}
}
