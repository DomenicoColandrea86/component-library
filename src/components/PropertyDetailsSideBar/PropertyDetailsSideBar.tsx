import React, { useState, ReactElement } from 'react';
import { b } from '../../utils/bem';
import { IBaseFields } from '../../models/base/fields';
import './PropertyDetailsSideBar.scss';

export const componentName = 'PropertyDetailsSideBar';

const block = b.with(componentName);

export enum LoanEventIds {
	Outstanding = 1,
	Watchlist = 2,
	SpecialServicing = 3,
	REO = 4,
	PaidOff = 5,
}

export enum LoanEventTypes {
	Watchlist = 'Watchlist',
	SpecialServicing = 'Special Service',
	REO = 'REO',
}

export interface IPropertyDetailsSideBarProps extends IBaseFields {
	isOpen?: boolean;
	loan?: any;
	handleClose?: () => void;
}

const PropertyDetailsSideBar: React.FC<IPropertyDetailsSideBarProps> = ({
	testId,
	isOpen,
	loan,
	handleClose,
}): ReactElement => {
	const [selectedTab, setSelectedTab] = useState<string | null>(
		// Temporarily turning this off as per George's request in CD-23161 - see comments in ticket
		// Once data has been corrected we will revert, for now default to `Watchlist`
		// loan ? LoanEventTypes[LoanEventIds[loan.loanEventId]] : null
		LoanEventTypes.Watchlist
	);

	const isDisabled: (type: string) => boolean = type => {
		let disabled = false;
		switch (type) {
			case LoanEventTypes.Watchlist:
				disabled = !(
					loan &&
					loan.watchlistComments &&
					!!loan.watchlistComments.length
				);
				break;
			case LoanEventTypes.SpecialServicing:
				disabled = !(
					loan &&
					loan.specialServicerComments &&
					!!loan.specialServicerComments.length
				);
				break;
			case LoanEventTypes.REO:
				disabled = !(loan && loan.reoComments && !!loan.reoComments.length);
				break;
			default:
				throw new Error('Invalid loan event type');
		}
		return disabled;
	};

	const renderHeader: () => string = () => {
		let header = '';

		if (loan) {
			switch (loan.loanEventId) {
				case LoanEventIds.Outstanding:
					header = 'Loan is current';
					break;
				case LoanEventIds.Watchlist:
					header = `Loan is on the ${loan.loanEvent}`;
					break;
				case LoanEventIds.SpecialServicing:
					header = `Loan is in ${loan.loanEvent}`;
					break;
				case LoanEventIds.REO:
					header = 'The lender has taken control of the property';
					break;
				default:
					throw new Error(`Invalid loan event id: ${loan.loanEventId}`);
			}
		}
		return header;
	};

	const getComments = (type, text) => (!text ? `No ${type} comments` : text);

	const renderComments: () => JSX.Element = () => {
		let troubledComments = '';
		if (selectedTab && loan) {
			switch (selectedTab) {
				case LoanEventTypes.Watchlist:
					troubledComments = getComments(
						LoanEventTypes.Watchlist,
						loan.watchlistComments
					);
					break;
				case LoanEventTypes.SpecialServicing:
					troubledComments = getComments(
						LoanEventTypes.SpecialServicing,
						loan.specialServicerComments
					);
					break;
				case LoanEventTypes.REO:
					troubledComments = getComments(LoanEventTypes.REO, loan.reoComments);
					break;
				default:
					throw new Error('Invalid loan event type');
			}
			return <p>{troubledComments}</p>;
		}
	};

	const renderBtns: JSX.Element[] = Object.keys(LoanEventTypes)
		.map(key => LoanEventTypes[key])
		.map(type => (
			<a
				key={type}
				data-testid={`${testId}-${type}-btn`}
				onClick={() => setSelectedTab(type)}
				className={block('btn', {
					active: selectedTab === type,
					// Temporarily removing this feature as per George's request in CD-23161 - see comments in ticket
					// Once data has been corrected we will revert, for now remove `disabled` state
					// disabled: isDisabled(type),
				})}
			>
				{type}
			</a>
		));

	return (
		<div className={block({ 'is-open': isOpen })} data-testid={testId}>
			<button
				data-testid={`${testId}-close-button`}
				className={block('close')}
				onClick={handleClose}
			></button>
			<div data-testid={`${testId}-content`} className={block('content')}>
				<h1 data-testid={`${testId}-header`} className={block('header')}>
					{renderHeader()}
				</h1>
				<div data-testid={`${testId}-btn-group`} className={block('btn-group')}>
					{renderBtns}
				</div>
				<div data-testid={`${testId}-comments`} className={block('comments')}>
					{renderComments()}
				</div>
			</div>
		</div>
	);
};

export default PropertyDetailsSideBar;
