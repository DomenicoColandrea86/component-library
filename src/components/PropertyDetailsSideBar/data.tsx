import { LoanEventIds, LoanEventTypes } from './PropertyDetailsSideBar';

export const Loan = {
	show: true,
	summary:
		'$7.5 m 1st Mortgage with <span class="name">BEAR 2006-PWR12</span> at  5.8% (Fixed) ',
	summaryLink:
		'$7.5 m 1st Mortgage with <a href="/#/lender-profile/69cb529b-298a-4263-b884-a7a653a7f363/summary"><span class="name">BEAR 2006-PWR12</span></a> at  5.8% (Fixed) ',
	loanStatus: 'Troubled',
	loanType: '1st Mortgage',
	qualifier: null,
	debtYield: '',
	totalLoanReserve: '',
	loanQualifier: '',
	loanAmount: '$7.5 m',
	loanId: 88636,
	interestRate: '5.8% Fixed',
	term: '120 Months',
	appraisal: '$9,398,496',
	originator: 'Wells Fargo',
	lender: 'C-III Asset Management OBO BEAR 2006-PWR12',
	lenderGroup: 'CMBS',
	mortgageBroker: null,
	loanAmortization: 'Partial Interest Only',
	originalLTV: '79.8%',
	loanDescription: '1.48x',
	origination: '06/01/2006',
	defeasance: '05/01/2009',
	prepayment: '02/01/2016',
	originalMaturity: '06/01/2016',
	comments: 'Trouble;',
	controlNumber: null,
	loanEventId: LoanEventIds.SpecialServicing,
	loanEvent: LoanEventTypes.SpecialServicing,
	loanEventDt: '09/20/2012',
	watchlistComments:
		'Subject is a retail property located in Tappahannock, VA. Q2 2012, occupancy is 77.99% and NCF DSCR is at 1.06x. \n\n06/04/2013: Property is on the watch list because NCF DSCR has dropped below threshold of 1.1x and occupancy is below 80%. The borrower is marketing the vacant space. Will continue to monitor property.',
	watchlistCommentsDt: '06/01/2013',
	specialServicerComments:
		"Loan transferred to special servicing 9/2012. Property condition reports show significant deferred maintenance (roofs and parking lot). Borrower indicated certain repairs were underway.  Borrower stated it will be unable to refinance the debt at matu     rity in 6/2016.  Borrower continues to remit Debt Service payments.  Borrower's request for a short sale to an unrelated third party has been approved by the Special Servicer.  Consummation of the transaction anticipated in 7/2016.",
	specialServicerCommentsDt: '06/01/2016',
	reoComments: null,
	reoCommentsDt: null,
	outstandingBalanceUSD_amt: '$6,593,401',
	outstandingBalanceDt: '2019-07-01T00:00:00',
	formatFinancingSummary: () => '',
	formatCurrency: () => '',
	formatInterest: () => '',
	showDate: () => false,
	formatDate: () => '',
	formatFinancingComments: () => '',
	formatLoanAmount: () => '',
	formatQualifier: () => '',
};
