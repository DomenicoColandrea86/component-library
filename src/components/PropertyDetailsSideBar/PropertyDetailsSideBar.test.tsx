import '@testing-library/jest-dom/extend-expect';
import * as React from 'react';
import PropertyDetailsSideBar, {
	LoanEventIds,
	LoanEventTypes,
	componentName,
	IPropertyDetailsSideBarProps,
} from './PropertyDetailsSideBar';
import { render, fireEvent, wait } from '@testing-library/react';
import { Loan } from './data';

describe('<PropertyDetailsSideBar />', () => {
	let props: IPropertyDetailsSideBarProps;
	beforeEach(() => {
		props = {
			testId: 'Test',
			isOpen: true,
			loan: Loan,
			handleClose: jest.fn(),
		};
	});

	const getCloseButton = queryByTestId =>
		queryByTestId(`${props.testId}-close-button`);
	const getContent = queryByTestId => queryByTestId(`${props.testId}-content`);
	const getHeader = queryByTestId => queryByTestId(`${props.testId}-header`);
	const getBtnGroup = queryByTestId =>
		queryByTestId(`${props.testId}-btn-group`);
	const getComments = queryByTestId =>
		queryByTestId(`${props.testId}-comments`);
	const getWatchListBtn = queryByTestId =>
		queryByTestId(`${props.testId}-${LoanEventTypes.Watchlist}-btn`);
	const getSpecialServicingBtn = queryByTestId =>
		queryByTestId(`${props.testId}-${LoanEventTypes.SpecialServicing}-btn`);
	const getREOBtn = queryByTestId =>
		queryByTestId(`${props.testId}-${LoanEventTypes.REO}-btn`);
	const mockFuncs = {
		formatFinancingSummary: jest.fn(),
		formatCurrency: jest.fn(),
		formatInterest: jest.fn(),
		showDate: jest.fn(),
		formatDate: jest.fn(),
		formatFinancingComments: jest.fn(),
		formatLoanAmount: jest.fn(),
		formatQualifier: jest.fn(),
	};
	const mockComment =
		'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.';
	const watchlistHeaderString = `Loan is on the ${LoanEventTypes.Watchlist}`;
	const specialServicingHeaderString = `Loan is in ${LoanEventTypes.SpecialServicing}`;
	const reoHeaderString = `The lender has taken control of the property`;

	it('should render correctly', () => {
		expect(render(<PropertyDetailsSideBar {...props} />)).toMatchSnapshot();
	});

	it('should render with the correct classes', () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(queryByTestId(props.testId)).toHaveClass(componentName);
		expect(getCloseButton(queryByTestId)).toHaveClass(
			`${componentName}__close`
		);
		expect(getContent(queryByTestId)).toHaveClass(`${componentName}__content`);
		expect(getHeader(queryByTestId)).toHaveClass(`${componentName}__header`);
		expect(getBtnGroup(queryByTestId)).toHaveClass(
			`${componentName}__btn-group`
		);
		expect(getWatchListBtn(queryByTestId)).toHaveClass(`${componentName}__btn`);
		expect(getSpecialServicingBtn(queryByTestId)).toHaveClass(
			`${componentName}__btn`
		);
		expect(getREOBtn(queryByTestId)).toHaveClass(`${componentName}__btn`);
		expect(getComments(queryByTestId)).toHaveClass(
			`${componentName}__comments`
		);
	});

	it('should set close button test id', () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(getCloseButton(queryByTestId)).toBeTruthy();
	});

	it('should set content test id', () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(getContent(queryByTestId)).toBeTruthy();
	});

	it('should set header test id', () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(getHeader(queryByTestId)).toBeTruthy();
	});

	it('should set button group test id', () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(getBtnGroup(queryByTestId)).toBeTruthy();
	});

	it('should set comments test id', () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(getComments(queryByTestId)).toBeTruthy();
	});

	it('should have correct modifier class', async () => {
		const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
		expect(queryByTestId(props.testId)).toHaveClass(
			`${componentName}--is-open`
		);
	});

	describe('close button', () => {
		it('should call handleClose', async () => {
			const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
			const button = getCloseButton(queryByTestId);
			fireEvent.click(button);
			await wait(() => expect(props.handleClose).toHaveBeenCalled());
		});
	});

	describe('watchlist', () => {
		beforeEach(() => {
			props.loan = {
				...Loan,
				...mockFuncs,
				loanEventId: LoanEventIds.Watchlist,
				loanEvent: LoanEventTypes.Watchlist,
				watchlistComments: mockComment,
				specialServicerComments: null,
				reoComments: null,
			};
		});

		describe('header', () => {
			it('should display correct string', async () => {
				const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
				expect(getHeader(queryByTestId)).toHaveTextContent(
					watchlistHeaderString
				);
			});
		});

		describe('btn-group', () => {
			it('should set correct modifier classes', async () => {
				const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
				expect(getWatchListBtn(queryByTestId)).toHaveClass(
					`${componentName}__btn--active`
				);
				expect(getSpecialServicingBtn(queryByTestId)).toHaveClass(
					`${componentName}__btn`
				);
				expect(getREOBtn(queryByTestId)).toHaveClass(`${componentName}__btn`);
			});
		});
	});

	describe('special servicing', () => {
		beforeEach(() => {
			props.loan = {
				...Loan,
				...mockFuncs,
				loanEventId: LoanEventIds.SpecialServicing,
				loanEvent: LoanEventTypes.SpecialServicing,
				watchlistComments: mockComment,
				specialServicerComments: mockComment,
				reoComments: null,
			};
		});

		describe('header', () => {
			it('should display correct string', async () => {
				const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
				expect(getHeader(queryByTestId)).toHaveTextContent(
					specialServicingHeaderString
				);
			});
		});

		describe('btn-group', () => {
			it('should set correct modifier classes', async () => {
				const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
				expect(getWatchListBtn(queryByTestId)).toHaveClass(
					`${componentName}__btn--active`
				);
				expect(getSpecialServicingBtn(queryByTestId)).toHaveClass(
					`${componentName}__btn`
				);
				expect(getREOBtn(queryByTestId)).toHaveClass(`${componentName}__btn`);
			});
		});
	});

	describe('reo', () => {
		beforeEach(() => {
			props.loan = {
				...Loan,
				...mockFuncs,
				loanEventId: LoanEventIds.REO,
				loanEvent: LoanEventTypes.REO,
				watchlistComments: mockComment,
				specialServicerComments: mockComment,
				reoComments: mockComment,
			};
		});

		describe('header', () => {
			it('should display correct string', async () => {
				const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
				expect(getHeader(queryByTestId)).toHaveTextContent(reoHeaderString);
			});
		});

		describe('btn-group', () => {
			it('should set correct modifier classes', async () => {
				const { queryByTestId } = render(<PropertyDetailsSideBar {...props} />);
				expect(getWatchListBtn(queryByTestId)).toHaveClass(
					`${componentName}__btn--active`
				);
				expect(getSpecialServicingBtn(queryByTestId)).toHaveClass(
					`${componentName}__btn`
				);
				expect(getREOBtn(queryByTestId)).toHaveClass(`${componentName}__btn`);
			});
		});
	});
});
