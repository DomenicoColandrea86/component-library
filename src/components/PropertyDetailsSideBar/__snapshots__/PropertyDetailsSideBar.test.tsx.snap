// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PropertyDetailsSideBar /> should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="PropertyDetailsSideBar PropertyDetailsSideBar--is-open"
        data-testid="Test"
      >
        <button
          class="PropertyDetailsSideBar__close"
          data-testid="Test-close-button"
        />
        <div
          class="PropertyDetailsSideBar__content"
          data-testid="Test-content"
        >
          <h1
            class="PropertyDetailsSideBar__header"
            data-testid="Test-header"
          >
            Loan is in Special Service
          </h1>
          <div
            class="PropertyDetailsSideBar__btn-group"
            data-testid="Test-btn-group"
          >
            <a
              class="PropertyDetailsSideBar__btn PropertyDetailsSideBar__btn--active"
              data-testid="Test-Watchlist-btn"
            >
              Watchlist
            </a>
            <a
              class="PropertyDetailsSideBar__btn"
              data-testid="Test-Special Service-btn"
            >
              Special Service
            </a>
            <a
              class="PropertyDetailsSideBar__btn"
              data-testid="Test-REO-btn"
            >
              REO
            </a>
          </div>
          <div
            class="PropertyDetailsSideBar__comments"
            data-testid="Test-comments"
          >
            <p>
              Subject is a retail property located in Tappahannock, VA. Q2 2012, occupancy is 77.99% and NCF DSCR is at 1.06x. 

06/04/2013: Property is on the watch list because NCF DSCR has dropped below threshold of 1.1x and occupancy is below 80%. The borrower is marketing the vacant space. Will continue to monitor property.
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="PropertyDetailsSideBar PropertyDetailsSideBar--is-open"
      data-testid="Test"
    >
      <button
        class="PropertyDetailsSideBar__close"
        data-testid="Test-close-button"
      />
      <div
        class="PropertyDetailsSideBar__content"
        data-testid="Test-content"
      >
        <h1
          class="PropertyDetailsSideBar__header"
          data-testid="Test-header"
        >
          Loan is in Special Service
        </h1>
        <div
          class="PropertyDetailsSideBar__btn-group"
          data-testid="Test-btn-group"
        >
          <a
            class="PropertyDetailsSideBar__btn PropertyDetailsSideBar__btn--active"
            data-testid="Test-Watchlist-btn"
          >
            Watchlist
          </a>
          <a
            class="PropertyDetailsSideBar__btn"
            data-testid="Test-Special Service-btn"
          >
            Special Service
          </a>
          <a
            class="PropertyDetailsSideBar__btn"
            data-testid="Test-REO-btn"
          >
            REO
          </a>
        </div>
        <div
          class="PropertyDetailsSideBar__comments"
          data-testid="Test-comments"
        >
          <p>
            Subject is a retail property located in Tappahannock, VA. Q2 2012, occupancy is 77.99% and NCF DSCR is at 1.06x. 

06/04/2013: Property is on the watch list because NCF DSCR has dropped below threshold of 1.1x and occupancy is below 80%. The borrower is marketing the vacant space. Will continue to monitor property.
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
